Multiple user interface (visual and functionalities) and backend updates

shtukas opened the pull request:
Multiple user interface and backend updates:

In this commit:

A. User Interface Visuals: the CSS directives for the Dashboard and the Editor have been updated. In particular the Editor has been entirely redesigned. 

B. User Interface Functionalities:
1. The dashboard has a correct rendering of the last modified time (possible due to an update in the data model, by which explainers now carry their last modified timestamp).
2. The dashboard now has a [Create New] button which allows for the creation of new items. Note that items are created with an id set to a UUID. A default title and default body are automatically set.
3. The Editor now has a "status bar" which indicates (when relevant) that an item is in draft state (meaning has underdone modifications since the last publish operation). This allows the user to be aware of the fact that the item current contents are not live yet. 
4. Along with the status bar comes a [Publish] button which performs the publishing of the current draft. 
5. URL Redirection from root `/` to `/explain`
6. Presence now reports the correct user initials

C. Backend Updates: The backend updates are mostly supporting the extension of the Explainer datatype and the additional operations. In particular we moved from 

```
case class Explainer(
              id                     : String,
              headline               : String,
              body                   : String)
```

to

```
case class ExplainerFacet(
               title             : String,
               body              : String,
               last_modified_time: Long)

case class ExplainerItem(
              id   : String,
              draft: ExplainerFacet,
              live : Option[ExplainerFacet])
```

shtukas committed:
removing misc empty lines

shtukas committed:
refactoring and update in HTML template

shtukas committed:
Update in Editor User Interface

shtukas committed:
-

shtukas committed:
Merge branch 'ph-20160715-02-backend-functions-updates' into pascal

shtukas committed:
Various Updates in Explainer Editor user interface

Notably
  - Complete rewrite of the CSS directives
  - Update in the HTML structure
  - Small updates in the scalatags

shtukas committed:
Merge branch 'ph-20160717-week-end-updates-01' into pascal

shtukas committed:
Enabling the creation of elements

- Expands the trait ExplainerApi with a function
  for the creation of a new explainer.

- Updates the User Interface with a working button to
  trigger the creation of a new explainer.

- Updated the backend code with support for the creation
  of a new explainer.

shtukas committed:
Enable redirection to explainer editor page after creation

Until now the user needed to reload the dashboard to update the listing
and then click on the newly created item. Now, this is no longer necessary.

shtukas committed:
Enabling dashboard ordering

This enables dashboard ordering. Displaying the items in decreasing last modification time,
where a modification is interpreted as an update of either the headline or the body.

shtukas committed:
Enabling dashboard last modification

This enables a user friendly display of the last modification datetime on the dashboard.

shtukas committed:
Status Bar

Enable informing the user that the current explainer data has been updated
(ie: is in a newer state than what has been last published).

shtukas commented:
CODE is currently running this branch

shtukas commented:
View of the dashboard:

<img width="1024" alt="picture 3" src="https://cloud.githubusercontent.com/assets/6035518/16902107/d756e48a-4c4c-11e6-8790-a9ff0a663b33.png">

shtukas commented:
View of the Editor: (with a status bar indicating draft state, and presence indicator -- both in red)

<img width="1026" alt="picture 2" src="https://cloud.githubusercontent.com/assets/6035518/16902111/e8fddcde-4c4c-11e6-9383-a9e09f512972.png">

philmcmahon commented on explainer-client/src/main/scala/ExplainEditorJS.scala:
```diff
@@ -45,74 +45,118 @@ object ExplainEditorJS {
-      Model.saveContent(explainerId, ExplainerUpdate("headline", headlineTag.value))
-      false
+      Model.updateFieldContent(explainerId, ExplainerUpdate("headline", headlineTag.value)).map { explainer: Explainer =>
+        republishStatusBar(explainer)
```
You could simplify this to 
`Model.updateFieldContent(explainerId, ExplainerUpdate("headline", headlineTag.value)).map(republishStatusBar)` - likewise below

shtukas replied:
Updated. Thanks !

shtukas committed:
small code refactoring

philmcmahon commented on explainer-server/app/controllers/ApiController.scala:
```diff
@@ -32,9 +32,23 @@ class ApiController @Inject() (val publicSettingsService: PublicSettingsService)
+    for {
+      explainer <- load(id)
+      _ <- update(id, "headline_published", explainer.headline)
+      _ <- update(id, "body_published"    , explainer.body)
```
This doesn't quite match up with what I had in my head of how publishing would work - is this just an intermediate thing before we set up the system we discussed on Friday?

shtukas replied:
Um... that's a good question. What we discussed on Friday had two parts. Draft state versus Publish state and the history of changes. For the latter @annebyrne indicated that there was no immediate need for the feature and could actually confuse the users (that being said it can always be added later). The problem now becomes how to distinguish between draft versus published, and having the Explainer object become as below was the simplest while neat solution.

```
case class Explainer(
              id                     : String,
              headline               : String,
              body                   : String,
              last_update_time_milli : Option[Long],
              headline_published     : Option[String],
              body_published         : Option[String])
```

In particular, this implies that the generation of the JSON file that is currently sent to Front will simply present `headline_published` and `body_published` as data (fields renamed to fit the current file grammar, of course).

shtukas replied:
As discussed will migrate to 

```
{
    id: String,
    draft: {
        headline : String,
        body     : String,
        last_update_time: Long 
    },
    live: {
        headline : String
        body     : String
        last_update_time: Long 
    }
}
```

```
case class ExplainerFacet(
    headline : String,       
    body     : String,
    last_modified_time: Long) 

case class ExplainerItem(
    id    : String,
    draft : ExplainerFacet        
    live  : Option(ExplainerFacet)) 
```

philmcmahon commented on explainer-server/app/models/ExplainerStore.scala:
```diff
@@ -41,16 +38,33 @@ object ExplainerStore {
   }
+
+  def create(): Future[Explainer] = {
+    val uuid = java.util.UUID.randomUUID.toString
```
What happens if there is a collision of IDs?

shtukas replied:
As discussed should not be a problem (it is safe to assume that a newly created UUID is different from any other UUID ever generated before).

philmcmahon commented on explainer-server/app/views/explainList.scala.html:
```diff
@@ -1,28 +1,36 @@
-                <div class="last_modified">Last modified</div>
-                <div class="created_by">Created by</div>
-                <div class="status">Status</div>
+                <div class="title"><b>Title</b></div>
```
you could give all these fields a 'heading' class then set the bold style (and anything else) on that

shtukas replied:
Done!

philmcmahon commented on explainer-client/src/main/scala/ExplainEditorJS.scala:
```diff
@@ -45,74 +45,112 @@ object ExplainEditorJS {
 
-      presenceClient.on("visitor-list-updated", { data: js.Object =>
-        val stateChange = upickle.default.read[StateChange](js.JSON.stringify(data))
+  def presenceEnvelop(explainerId: String ,area: String, field: Element) = {
```
I'm not sure this is the most understandable name. Is this the function that adds the presence indicator? Maybe the old function name could work better. But I'm sure you had a good reason for changing it!

shtukas replied:
Will revert to the old name actually, now that you mention it...

shtukas replied:
Actually I went for "presenceWrapper".

shtukas committed:
using css directive for bold headers

shtukas committed:
renaming presenceEnvelop into presenceWrapper

shtukas committed:
Change in Data Model

In this commit, we move from

```
case class Explainer(
              id                     : String,
              headline               : String,
              body                   : String,
              last_update_time_milli : Option[Long],
              headline_published     : Option[String],
              body_published         : Option[String])
```

to

```
case class ExplainerFacet(
               title             : String,
               body              : String,
               last_modified_time: Long)

case class ExplainerItem(
              id   : String,
              draft: ExplainerFacet,
              live : Option[ExplainerFacet])
```

which essentially defines an explainer as carrying two
objects of the same shape.

Other updates:
1. "headline" has been renamed to "title"
2. The CSS directives suffix "envelop" has been replaced by "wrapper"

shtukas committed:
CSS directives update in ExplainEditorJS

... had forgotten them :)

shtukas committed:
Redirection from root to /explain

shtukas commented:
This should not prevent the PR to be accepted, but for the record I am seeing those errors on CODE, but not on local.

![picture 1](https://cloud.githubusercontent.com/assets/6035518/16931950/8b110c82-4d3b-11e6-91d7-65e35a3b9e3e.png)

philmcmahon commented on explainer-client/src/main/scala/ExplainEditorJS.scala:
```diff
@@ -37,86 +37,135 @@ object Ajaxer extends autowire.Client[Js.Value, Reader, Writer]{
-    placeholder:="headline",
-    autofocus:=true
-  )
+  def presenceWrapper(explainerId: String ,area: String, field: Element,presenceClient: PresenceClient) = {
```
Didn't we agree to just leave this with the old name? turnOnPresenceFor

shtukas replied:
Yes, but then you explain to me that "wrapper" is what is usually used in those cases :)

philmcmahon replied:
Yeah but in this particular case, this function is literally 'turning on presence' for that block so the old name seems to make more sense.

philmcmahon commented:
Sorry to be awkward, but could you move the presence changes into a different pull request? It's getting a bit difficult to review

philmcmahon commented:
(just the last 2 commits, not the stuff you did over the weekend)

philmcmahon commented on explainer-client/src/main/scala/ExplainEditorJS.scala:
```diff
@@ -37,86 +37,135 @@ object Ajaxer extends autowire.Client[Js.Value, Reader, Writer]{
+    }else{
+      ""
+    }
+    div(id:="explainer-editor__ops-wrapper__status-bar-env__status-bar",cls:="red")(status)
```
Tiny thing but explainer-editor__ops-wrapper__status-bar-_wrapper_

shtukas commented:
The branch is back to commit 17a0d6a78773322695be0a8207d22788c3d1d179 , but I need to add a non presence update

shtukas committed:
Update in page title

shtukas commented:
Ok, branch is up to date without presence related change. Will leave it as it is. Cheers,

shtukas commented:
Available on CODE

philmcmahon commented:
With these changes:
1. use `turningPresenceFor`.
2. In 
   explainer-editor__ops-wrapper__status-bar-env__status-bar
   explainer-editor__ops-wrapper__status-bar-wrapper

👍

shtukas committed:
Renaming presenceWrapper back to turnOnPresenceFor

shtukas committed:
Renaming explainer-editor__ops-wrapper__status-bar-env into explainer-editor__ops-wrapper__status-bar-wrapper

shtukas commented:
Latest changes on CODE.

philmcmahon commented:
👍
